File: examples/starters/langgraph/starter.py
"""
ACP Starter — LangGraph (Python)

The minimum code to wire ACP policy & audit into a LangChain 1.x /
LangGraph agent. Copy this folder, swap the placeholder tool for your
real one, ship.

Pattern: middleware (langchain >= 1.3.3). `ACPMiddleware()` registers on
`create_agent`'s middleware stack; its tool-call wrap hook governs EVERY
tool — no per-function decorators. Denials mean the tool never runs; the
model receives "tool_error: <reason>" as the tool result and adapts.
Redacted outputs replace the original before the model sees them.

Run:  bash run.sh
"""

from __future__ import annotations

import json
import os
from pathlib import Path

from dotenv import load_dotenv

from acp_langchain import ACPMiddleware, configure, set_context
from langchain.agents import create_agent
from langchain.tools import tool

# ── 1. Load ACP credentials from ./.env next to this file.
load_dotenv(Path(__file__).parent / ".env")
for v in ("OPENAI_API_KEY", "ACP_USER_TOKEN"):
    if not os.environ.get(v):
        raise SystemExit(f"Missing {v} — copy .env.example → .env")

# ── 2. Point the SDK at your ACP gateway. Once per process.
configure(
    base_url=os.environ.get("ACP_GATEWAY_URL", "https://api.agenticcontrolplane.com"),
    client_header="acp-langgraph-starter/0.2.0",
)


# ── 3. Your tool. Just `@tool` — ACPMiddleware below already covers it.
# REPLACE the body with your real logic (DB lookup, API call, etc.).
@tool
def lookup_record(id: str) -> str:
    """Look up a record by ID. Replace with your real tool description."""
    return json.dumps({"id": id, "status": "placeholder", "note": "replace me"})


def main() -> None:
    # ── 4. Bind identity for every policy call inside. Without this
    # set_context call, ACP pre/post hooks silently no-op.
    set_context(
        user_token=os.environ["ACP_USER_TOKEN"],
        agent_name="my-langgraph-agent",  # Rename for your agent.
        agent_tier="background",
    )

    # ── 5. Build the agent. `create_agent` is the LangChain 1.x idiom
    # (replaces the legacy `langgraph.prebuilt.create_react_agent`).
    # One middleware line governs every tool — the one above and any
    # you add later.
    agent = create_agent(
        model="openai:gpt-4o-mini",
        tools=[lookup_record],
        middleware=[ACPMiddleware()],
    )

    result = agent.invoke(
        {"messages": [{"role": "user", "content": "Look up record id=abc-123 and tell me what you find."}]}
    )
    # The agent returns a dict with `messages`; the last message is the
    # model's final answer.
    print(result["messages"][-1].content)


if __name__ == "__main__":
    main()

File: examples/starters/claude-agent-sdk/starter.ts
/**
 * ACP Starter — Claude Agent SDK (TypeScript)
 *
 * The minimum code needed to run a Claude-powered agent with ACP governance
 * on every tool call. Copy this folder, swap the placeholder tool for your
 * real one, and ship.
 *
 * Governance pattern: decorator. `governHandlers({...})` wraps each tool so
 * every call runs `preToolUse → handler → postToolOutput`. Denials return
 * a `"tool_error: <reason>"` string — the model sees it as a tool result
 * and adapts.
 *
 * Run:   bash run.sh
 */

import dotenv from "dotenv";
import path from "node:path";
import { fileURLToPath } from "node:url";

import Anthropic from "@anthropic-ai/sdk";
import {
  configure,
  governHandlers,
  withContext,
} from "@agenticcontrolplane/governance-anthropic";

// ── 1. Load ACP credentials from ./.env next to this file.
//   Copy .env.example → .env and fill in your keys.
const __dirname = path.dirname(fileURLToPath(import.meta.url));
dotenv.config({ path: path.join(__dirname, ".env") });

const { ANTHROPIC_API_KEY, ACP_USER_TOKEN, ACP_GATEWAY_URL = "https://api.agenticcontrolplane.com" } = process.env;
if (!ANTHROPIC_API_KEY) throw new Error("Missing ANTHROPIC_API_KEY");
if (!ACP_USER_TOKEN) throw new Error("Missing ACP_USER_TOKEN (gsk_...)");

// ── 2. Point the governance SDK at your ACP gateway. Once per process.
configure({ baseUrl: ACP_GATEWAY_URL });

const anthropic = new Anthropic({ apiKey: ANTHROPIC_API_KEY });

// ── 3. Your tool handlers. `governHandlers` wraps each one with the ACP
// pre/post hook protocol — every call is checked before it runs and audited
// after. Replace the body of `lookup_record` with your actual tool logic.
// Add more handlers to the map as needed.
const handlers = governHandlers({
  lookup_record: async ({ id }: { id: string }) => {
    // REPLACE THIS with your real tool (DB lookup, API call, file read, etc.).
    return { id, status: "placeholder", note: "replace me with real logic" };
  },
});

// ── 4. Tool definitions passed to the Messages API.
const tools: Anthropic.Tool[] = [
  {
    name: "lookup_record",
    description: "Look up a record by ID. Replace with your real tool description.",
    input_schema: {
      type: "object",
      properties: { id: { type: "string", description: "The record ID to look up." } },
      required: ["id"],
    },
  },
];

// ── 5. Agent loop. Standard Anthropic tool-use pattern — nothing bespoke.
async function run(userMessage: string) {
  const messages: Anthropic.MessageParam[] = [{ role: "user", content: userMessage }];

  for (let i = 0; i < 10; i++) {
    const resp = await anthropic.messages.create({
      model: "claude-sonnet-4-6",
      max_tokens: 4096,
      tools,
      messages,
    });
    messages.push({ role: "assistant", content: resp.content });

    if (resp.stop_reason === "end_turn") {
      for (const b of resp.content) if (b.type === "text") console.log(b.text);
      return;
    }
    if (resp.stop_reason !== "tool_use") return;

    const results: Anthropic.ToolResultBlockParam[] = [];
    for (const block of resp.content) {
      if (block.type !== "tool_use") continue;
      const fn = handlers[block.name as keyof typeof handlers];
      const out = fn ? await fn(block.input as any) : `tool_error: unknown tool ${block.name}`;
      results.push({
        type: "tool_result",
        tool_use_id: block.id,
        content: typeof out === "string" ? out : JSON.stringify(out),
      });
    }
    messages.push({ role: "user", content: results });
  }
}

// ── 6. `withContext` binds the user's identity for every governance call
// inside. Without it, pre/post hooks silently no-op — the agent runs but
// nothing is governed. Always wrap your request handler body with this.
async function main() {
  await withContext(
    {
      userToken: ACP_USER_TOKEN!,
      agentName: "my-agent", // Rename to match your agent.
      agentTier: "background", // or "interactive" / "subagent" / "api"
    },
    () => run("Look up record id=abc-123 and tell me what you find."),
  );
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
